# Data Cleaning and Transformation
## Notebook 1

Group project by:
    - Tinuke Durotolu
    - Sara Ruini
    - Susan Yousefi


<b> Data source: </b> https://archive.ics.uci.edu/ml/datasets/Diabetes+130-US+hospitals+for+years+1999-2008# 

<b> Reference: </b>
Beata Strack, Jonathan P. DeShazo, Chris Gennings, Juan L. Olmo, Sebastian Ventura, Krzysztof J. Cios, and John N. Clore, “Impact of HbA1c Measurement on Hospital Readmission Rates: Analysis of 70,000 Clinical Database Patient Records,” BioMed Research International, vol. 2014, Article ID 781670, 11 pages, 2014. 

In [1]:
#libraries needed
import pandas as pd
import numpy as np
import os

In [2]:
df = pd.read_csv("diabetic_data.csv", encoding="Latin-1")

## Visualising the datasets

First of all, we are going to visualise the datatype of each column in the dataset and see if there is any missing values.

Our aim is to remove unnecessary values and columns and remove any information that we consider redundant.

## Removing unnecessary values

We will visualise the info of the dataset (presence of null-values) for each column, show the head and tail of the dataset and show the percentage of unique value_counts. 

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
encounter_id                101766 non-null int64
patient_nbr                 101766 non-null int64
race                        101766 non-null object
gender                      101766 non-null object
age                         101766 non-null object
weight                      101766 non-null object
admission_type_id           101766 non-null int64
discharge_disposition_id    101766 non-null int64
admission_source_id         101766 non-null int64
time_in_hospital            101766 non-null int64
payer_code                  101766 non-null object
medical_specialty           101766 non-null object
num_lab_procedures          101766 non-null int64
num_procedures              101766 non-null int64
num_medications             101766 non-null int64
number_outpatient           101766 non-null int64
number_emergency            101766 non-null int64
number_inpatient            10176

There are a total of 101766 data entries in this dataset.

Before analysing this dataframe further, we will need to replace the NaN values with 'missing',

In [4]:
to_drop = ['weight','medical_specialty', 'admission_type_id', 'discharge_disposition_id','max_glu_serum', 'admission_source_id', 'encounter_id','patient_nbr', 'diag_1', 'diag_2', 'diag_3', 'payer_code', 'A1Cresult']
df.drop(to_drop, inplace=True, axis=1)
all_cols = list(df.columns)

In [5]:
for col in all_cols:
    df[col] = df[col].fillna('missing')
df.head(20)

race  gender       age  time_in_hospital  num_lab_procedures  \
0         Caucasian  Female    [0-10)                 1                  41   
1         Caucasian  Female   [10-20)                 3                  59   
2   AfricanAmerican  Female   [20-30)                 2                  11   
3         Caucasian    Male   [30-40)                 2                  44   
4         Caucasian    Male   [40-50)                 1                  51   
5         Caucasian    Male   [50-60)                 3                  31   
6         Caucasian    Male   [60-70)                 4                  70   
7         Caucasian    Male   [70-80)                 5                  73   
8         Caucasian  Female   [80-90)                13                  68   
9         Caucasian  Female  [90-100)                12                  33   
10  AfricanAmerican  Female   [40-50)                 9                  47   
11  AfricanAmerican    Male   [60-70)                 7                  62   
12        Caucasian  Female   [40-50)                 7                  60   
13        Caucasian    Male   [80-90)                10                  55   
14  AfricanAmerican  Female   [60-70)                 1                  49   
15  AfricanAmerican    Male   [60-70)                12                  75   
16  AfricanAmerican    Male   [50-60)                 4                  45   
17        Caucasian  Female   [50-60)                 3                  29   
18  AfricanAmerican    Male   [70-80)                 5                  35   
19                ?    Male   [70-80)                 6                  42   

    num_procedures  num_medications  number_outpatient  number_emergency  \
0                0                1                  0                 0   
1                0               18                  0                 0   
2                5               13                  2                 0   
3                1               16                  0                 0   
4                0                8                  0                 0   
5                6               16                  0                 0   
6                1               21                  0                 0   
7                0               12                  0                 0   
8                2               28                  0                 0   
9                3               18                  0                 0   
10               2               17                  0                 0   
11               0               11                  0                 0   
12               0               15                  0                 1   
13               1               31                  0                 0   
14               5                2                  0                 0   
15               5               13                  0                 0   
16               4               17                  0                 0   
17               0               11                  0                 0   
18               5               23                  0                 0   
19               2               23                  0                 0   

    number_inpatient  ...  citoglipton insulin glyburide-metformin  \
0                  0  ...           No      No                  No   
1                  0  ...           No      Up                  No   
2                  1  ...           No      No                  No   
3                  0  ...           No      Up                  No   
4                  0  ...           No  Steady                  No   
5                  0  ...           No  Steady                  No   
6                  0  ...           No  Steady                  No   
7                  0  ...           No      No                  No   
8                  0  ...           No  Steady                  No   
9                  0  ...           No  Steady            

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 37 columns):
race                        101766 non-null object
gender                      101766 non-null object
age                         101766 non-null object
time_in_hospital            101766 non-null int64
num_lab_procedures          101766 non-null int64
num_procedures              101766 non-null int64
num_medications             101766 non-null int64
number_outpatient           101766 non-null int64
number_emergency            101766 non-null int64
number_inpatient            101766 non-null int64
number_diagnoses            101766 non-null int64
metformin                   101766 non-null object
repaglinide                 101766 non-null object
nateglinide                 101766 non-null object
chlorpropamide              101766 non-null object
glimepiride                 101766 non-null object
acetohexamide               101766 non-null object
glipizide                   10

In [7]:
df.describe(include='all')

race  gender      age  time_in_hospital  num_lab_procedures  \
count      101766  101766   101766     101766.000000       101766.000000   
unique          6       3       10               NaN                 NaN   
top     Caucasian  Female  [70-80)               NaN                 NaN   
freq        76099   54708    26068               NaN                 NaN   
mean          NaN     NaN      NaN          4.395987           43.095641   
std           NaN     NaN      NaN          2.985108           19.674362   
min           NaN     NaN      NaN          1.000000            1.000000   
25%           NaN     NaN      NaN          2.000000           31.000000   
50%           NaN     NaN      NaN          4.000000           44.000000   
75%           NaN     NaN      NaN          6.000000           57.000000   
max           NaN     NaN      NaN         14.000000          132.000000   

        num_procedures  num_medications  number_outpatient  number_emergency  \
count    101766.000000    101766.000000      101766.000000     101766.000000   
unique             NaN              NaN                NaN               NaN   
top                NaN              NaN                NaN               NaN   
freq               NaN              NaN                NaN               NaN   
mean          1.339730        16.021844           0.369357          0.197836   
std           1.705807         8.127566           1.267265          0.930472   
min           0.000000         1.000000           0.000000          0.000000   
25%           0.000000        10.000000           0.000000          0.000000   
50%           1.000000        15.000000           0.000000          0.000000   
75%           2.000000        20.000000           0.000000          0.000000   
max           6.000000        81.000000          42.000000         76.000000   

        number_inpatient  ...  citoglipton insulin glyburide-metformin  \
count      101766.000000  ...       101766  101766              101766   
unique               NaN  ...            1       4                   4   
top                  NaN  ...           No      No                  No   
freq                 NaN  ...       101766   47383              101060   
mean            0.635566  ...          NaN     NaN                 NaN   
std             1.262863  ...          NaN     NaN                 NaN   
min             0.000000  ...          NaN     NaN                 NaN   
25%             0.000000  ...          NaN     NaN                 NaN   
50%             0.000000  ...          NaN     NaN                 NaN   
75%             1.000000  ...          NaN     NaN                 NaN   
max            21.000000  ...          NaN     NaN                 NaN   

       glipizide-metformin glimepiride-pioglitazone metformin-rosiglitazone  \
count               101766                   101766                  101766   
unique                   2                        2                       2   
top                     No                       No                      No   
freq                101753                   101765                  101764   
mean                   NaN                      NaN                     NaN   
std                    NaN                      NaN                     NaN   
min                    NaN                      NaN                     NaN   
25%                    NaN                      NaN                     NaN   
50%                    NaN                      NaN                     NaN   
75%                    NaN                      NaN                     NaN   
max                    NaN                      NaN                     NaN   

       metformin-pioglitazone  change diabetesMed readmitted  
count                  101766  101766      101766     101766  
unique                      2       2           2          3  
top                        No      No         Yes         NO  
freq                   101765   54755       78363

In [8]:
def value_counter(dataset):
    for index in df.columns:
        print("---- Index: " + index + "----")
        print(df[index].value_counts(normalize=True, sort=True))
value_counter(df)

---- Index: race----
Caucasian          0.747784
AfricanAmerican    0.188766
?                  0.022336
Hispanic           0.020017
Other              0.014799
Asian              0.006299
Name: race, dtype: float64
---- Index: gender----
Female             0.537586
Male               0.462384
Unknown/Invalid    0.000029
Name: gender, dtype: float64
---- Index: age----
[70-80)     0.256156
[60-70)     0.220928
[50-60)     0.169565
[80-90)     0.168986
[40-50)     0.095169
[30-40)     0.037095
[90-100)    0.027445
[20-30)     0.016282
[10-20)     0.006790
[0-10)      0.001582
Name: age, dtype: float64
---- Index: time_in_hospital----
3     0.174479
2     0.169251
1     0.139614
4     0.136824
5     0.097931
6     0.074082
7     0.057573
8     0.043148
9     0.029499
10    0.023014
11    0.018228
12    0.014229
13    0.011890
14    0.010239
Name: time_in_hospital, dtype: float64
---- Index: num_lab_procedures----
1      0.031523
43     0.027553
44     0.024527
45     0.023348
38     0.02

Some columns have a lot of unique numeric values which could be put in ranges. Let's display them to make each category more readable.  These columns are: 'num_lab_procedures', 'num_medications', 

In [9]:
cols_to_range = ['num_lab_procedures', 'num_medications']
for col in cols_to_range:
    unique_val_col = df[col].unique().shape
    print("Column: '" + col + "'. Number of unique values: " + str(unique_val_col) + ".")
    print(df[col].value_counts(bins=10, sort=True))
    print("-------")

Column: 'num_lab_procedures'. Number of unique values: (118,).
(40.3, 53.4]      27814
(27.2, 40.3]      21826
(53.4, 66.5]      20558
(0.868, 14.1]     10318
(14.1, 27.2]      10141
(66.5, 79.6]       8876
(79.6, 92.7]       1915
(92.7, 105.8]       290
(105.8, 118.9]       23
(118.9, 132.0]        5
Name: num_lab_procedures, dtype: int64
-------
Column: 'num_medications'. Number of unique values: (75,).
(9.0, 17.0]     45079
(17.0, 25.0]    24813
(0.919, 9.0]    20515
(25.0, 33.0]     7961
(33.0, 41.0]     2183
(41.0, 49.0]      730
(49.0, 57.0]      320
(57.0, 65.0]      131
(65.0, 73.0]       29
(73.0, 81.0]        5
Name: num_medications, dtype: int64
-------


## Data Modification

### Removing missing/unclear and null values.


In [10]:
df = df[df['gender']!='Unknown/Invalid']
df = df[df['race']!='?']

### Renaming columns

In [11]:
df.columns

Index(['race', 'gender', 'age', 'time_in_hospital', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'number_diagnoses', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

The goal is to:
    - Clarify what the column in the data is;
    - Conform to python standards (e.g. using snakecase)

In [12]:
new_col_names = {
    "time_in_hospital" : "days_in_hospital",
    "admission_type_id": "admission_type",
    "discharge_disposition_id": "discharge_disposition_type",
    "num_procedures": "num_not_lab_procedures",
    "num_medications": "num_current_medications",
    "num_diagnoses": "num_existing_conditions",
    "change": "change_in_meds",
    "diabetesMed": "diabates_med_prescribed",
    "number_emergency": "num_previous_emergencies",
    "number_outpatient": "num_outpatient_appointments",
    "number_inpatient": "num_inpatient_overnight_stays"
    }
df = df.rename(columns=new_col_names)
df.columns

Index(['race', 'gender', 'age', 'days_in_hospital', 'num_lab_procedures',
       'num_not_lab_procedures', 'num_current_medications',
       'num_outpatient_appointments', 'num_previous_emergencies',
       'num_inpatient_overnight_stays', 'number_diagnoses', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change_in_meds', 'diabates_med_prescribed',
       'readmitted'],
      dtype='object')

### Renaming values for each medicine

In [13]:
list_of_meds = ['metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone']

new_values = {"Up": "dosage_increased", "Down": "dosage_decreased","Steady":"no_change_dosage", "No":"not_prescribed"}
for medicine in list_of_meds:
    df[medicine] = df[medicine].replace(new_values)

for col in list_of_meds[:3]:
    print("----" + col + "----")
    print(df[col].value_counts(sort=True))
    print("-------")

----metformin----
not_prescribed      79971
no_change_dosage    17920
dosage_increased     1039
dosage_decreased      562
Name: metformin, dtype: int64
-------
----repaglinide----
not_prescribed      97963
no_change_dosage     1375
dosage_increased      109
dosage_decreased       45
Name: repaglinide, dtype: int64
-------
----nateglinide----
not_prescribed      98796
no_change_dosage      662
dosage_increased       23
dosage_decreased       11
Name: nateglinide, dtype: int64
-------


In [14]:
df.describe(include='all')

race  gender      age  days_in_hospital  num_lab_procedures  \
count       99492   99492    99492      99492.000000        99492.000000   
unique          5       2       10               NaN                 NaN   
top     Caucasian  Female  [70-80)               NaN                 NaN   
freq        76099   53575    25468               NaN                 NaN   
mean          NaN     NaN      NaN          4.398454           43.072971   
std           NaN     NaN      NaN          2.986972           19.695587   
min           NaN     NaN      NaN          1.000000            1.000000   
25%           NaN     NaN      NaN          2.000000           31.000000   
50%           NaN     NaN      NaN          4.000000           44.000000   
75%           NaN     NaN      NaN          6.000000           57.000000   
max           NaN     NaN      NaN         14.000000          132.000000   

        num_not_lab_procedures  num_current_medications  \
count             99492.000000             99492.000000   
unique                     NaN                      NaN   
top                        NaN                      NaN   
freq                       NaN                      NaN   
mean                  1.340540                16.026545   
std                   1.703686                 8.119809   
min                   0.000000                 1.000000   
25%                   0.000000                10.000000   
50%                   1.000000                15.000000   
75%                   2.000000                20.000000   
max                   6.000000                81.000000   

        num_outpatient_appointments  num_previous_emergencies  \
count                  99492.000000              99492.000000   
unique                          NaN                       NaN   
top                             NaN                       NaN   
freq                            NaN                       NaN   
mean                       0.373357                  0.201343   
std                        1.276805                  0.939981   
min                        0.000000                  0.000000   
25%                        0.000000                  0.000000   
50%                        0.000000                  0.000000   
75%                        0.000000                  0.000000   
max                       42.000000                 76.000000   

        num_inpatient_overnight_stays  ...     citoglipton         insulin  \
count                    99492.000000  ...           99492           99492   
unique                            NaN  ...               1               4   
top                               NaN  ...  not_prescribed  not_prescribed   
freq                              NaN  ...           99492           46495   
mean                         0.643107  ...             NaN             NaN   
std                          1.271370  ...             NaN             NaN   
min                          0.000000  ...             NaN             NaN   
25%                          0.000000  ...             NaN             NaN   
50%                          0.000000  ...             NaN             NaN   
75%                          1.000000  ...             NaN             NaN   
max                         21.000000  ...             NaN             NaN   

       glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
count                99492               99492                    99492   
unique                   4                   2                        2   
top         not_prescribed      not_prescribed           not_prescribed   
freq                 98812               99479                    99491   
mean                   NaN                 NaN                      NaN   
std                    NaN                 NaN                      NaN   
min                    NaN                 NaN                      NaN   
25%                    NaN                 NaN                      NaN  

In [15]:
df.to_csv("cleaned_data_bn.csv")